### ISSUE TITLE
CUDA memory leak for pytorch nn.Parameter 

### ISSUE THREAD

**Issue by ueser** [dmlc/dgl#698]
## 🐛 CUDA leaks memory for pytorch nn.Parameter fed as node features. 
Eventually, throws an OOM error. 

## To Reproduce

minimum example code to reproduce the behaviour:
```
g = dgl.DGLGraph()
g.add_nodes(10000)
g.ndata['x'] = nn.Parameter(torch.randn(10000,50)).cuda()

def assign_feature(nodes): return {'y': nodes.data['x']}
for i in range(100000):
    g.apply_nodes(func=assign_feature, v=np.arange(5000, 10000))
```
Then watch the cuda memory from terminal:
`$ nvidia-smi -l`

## Expected behavior

The memory-usage ncreases at every iteration. On the other hand, if you set the feature data a non-parameter tensor, e.g.
`g.ndata['x'] = torch.randn(10000,50).cuda()`

everything seems to be fine. 

## Environment

 - DGL Version : '0.3'
 - Backend Library & Version (e.g., PyTorch 0.4.1, MXNet/Gluon 1.3): PyTorch 1.1.0
 - OS: Ubuntu
 - How you installed DGL (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version (if applicable): 10.0
 - GPU models and configuration (e.g. V100): Tesla M60

**Comment by VoVAllen** [comment]
Hi,

If you change `assign_feature` to `def assign_feature(nodes): return {'y': nodes.data['x'].detach()}`, you would fine everything works fine.

The reason why here memory leak occurs is due to pytorch will keep tracing tensor with partial set operation. For example, the operation here can be considered as 
```python
for i in range(1000000):
  g.ndata['y'][_updated_nodes_idx] = _results_from_apply_func
```
Each time it will create a new tensor. And all previous tensors are still kept for backward tracing. Every `y` holds a reference on `x` for further backward computation. That's why memory keep increasing. If you changed it to non-parameter or detach it, the reference from `y` to `x` would no longer exists.

**Comment by ueser** [comment]
But then how would you backprop if you detach? 

Seems to be the `apply_nodes` or `apply_edges` functions are limited to the whole set operations to avoid memory leakage? 

Is there an obvious alternative that I'm missing here? 

Right now, I'm explicitly deleting the features from the entire node set before a new partial set assignment. But it is not ideal.

**Comment by VoVAllen** [comment]
You can manually delete previous tensor if you don't want them. Code below should work.
```python
g = dgl.DGLGraph() 
g.add_nodes(10000) 
g.ndata['x'] = nn.Parameter(torch.randn(10000,50)).cuda() 
from functools import partial 
def assign_feature(nodes, i): return {f'y_{i}': nodes.data['x']} 
for i in range(100000): 
    g.apply_nodes(func=partial(assign_feature,i=i), v=np.arange(5000, 10000)) 
    if i>0: del g.ndata[f'y_{i-1}'] 
```

**Comment by VoVAllen** [comment]
I just saw you said you deleted the features from entire node set. Why is this not ideal?

**Comment by ueser** [comment]
> I just saw you said you deleted the features from entire node set. Why is this not ideal?

Well, first of all, you need to delete the feature from all of the nodes in order to update the partial. 
Say you need to update different sets of nodes (or edges) with different functions, or not in one-go (this might actually be a re-appearing pattern in heterogeneous networks (hetnet)). Then, either you delete the feature right before you update a partial set and you will lose the other set of node's information, or you overwrite partially and it will create unreferenced garbages which won't be traced via garbage collector. 

That's why I think it is not ideal. 
Let me know if I'm wrong.

**Comment by VoVAllen** [comment]
We are developing the support for heterogeneous graph, which store different type of node in different tensors. Would this solve your problem?

**Comment by jermainewang** [comment]
This looks like a bug on our side. Ideally, the example DGL codes should be equivalent to the following torch codes:
```python
import torch as th
import torch.nn as nn
import numpy as np

x = nn.Parameter(th.randn(10000, 50)).cuda()
y = th.zeros((10000, 50)).cuda()
v = th.tensor(np.arange(5000, 10000)).cuda()
for i in range(100000):
    print(i)
    data = th.index_select(x, 0, v)
    y = y.index_copy(0, v, data)
```
, which does not run OOM.

**Comment by jermainewang** [comment]
Ok, I could reproduce the problem in pure torch codes:
```python
import torch as th
import torch.nn as nn
import numpy as np

x = nn.Parameter(th.randn(10000, 50)).cuda()
y = th.zeros((10000, 50)).cuda()
for i in range(100000):
    print(i)
    v = th.tensor(np.arange(5000, 10000)).cuda()
    data = th.index_select(x, 0, v)
    y = y.index_copy(0, v, data)
```
You could see that the difference is the `v` index tensor is now a local variable within the loop. In this case, the computation graph will capture the temporary `v` tensor in each iteration (in order to do backward propagation). Although they are all the same tensor, pytorch can now way understand that.

In terms of DGL, `g.apply_nodes(func=assign_feature, v=np.arange(5000, 10000))` will create a `v` tensor on cuda to index the feature tensor. Since a numpy array is provided, DGL will create a temporary tensor each time so causing the problem.

Changing your code to following will solve the issue:
```python
import dgl
import torch
import torch.nn as nn
import numpy as np

g = dgl.DGLGraph()
g.add_nodes(10000)
g.ndata['x'] = nn.Parameter(torch.randn(10000,50)).cuda()
# NOTE: the v tensor is outside of the loop
v = torch.tensor(np.arange(5000, 10000)).cuda()

def assign_feature(nodes): return {'y': nodes.data['x']}
for i in range(100000):
    print(i)
    g.apply_nodes(func=assign_feature, v=v)
```

Overall, I feel this is definitely a pitfall that should be aware of, but it is not a DGL bug. @ueser Let me know if you have any good idea to prevent others to fall for this again.

**Comment by ueser** [comment]
@jermainewang 
yes, out of loop index def works.  But it still presents a limitation for dynamic index selection. 
When you put v inside the for loop, it gives an OOM even if you explicitly delete the indices after apply_nodes. 

@VoVAllen 
Yes, I hope the next hetnet capability solves the gpu issues. 

But meanwhile, I hacked together a solution for my heterogenous network processing and now I don't have any GPU leakage at all. First,  I introduced a dict-like data structure so that I can delete the data partially:
```python
class Package(dict):
    def __init__(self, node_indices, *args, **kwargs):
        super(Package, self).__init__(*args, **kwargs)
        self.indices = node_indices
        
    def __setitem__(self, key, value):
        if key in self.keys():
            del self[key]
        return super(Package, self).__setitem__(key, value)
```
Then I move all my `apply_nodes` operations outside of the `dgl` using `pacakge`. When I need edge operations, I shuttle the data into my `dgl` graph:

```python
def deliver_to_graph(package, fields=None):
       '''
        Transfers data for subset of nodes to the graph. 
        
        Parameters
        -----------
         package : Package
              a dict-like package object that holds node indices and the associated data
        '''
        if fields is None:
           fields = package.keys()

        for key in fields:
           graph.nodes[package.indices].data[key] = package[key]
```

Of course with an accompanying clean up function:

```python
def flush(keep=None):
        '''
        Explicitly deletes the tensors from nodes and edges to prevent orphan parameter accumulation.
        
        Parameters
        -----------
         keep : list
               list of fields of ndata or edata to keep
        '''
        key_list = list(graph.ndata.keys())
        for key in  key_list:
            if key in keep:
                continue
            del graph.ndata[key]

        key_list = list(graph.edata.keys())
        for key in key_list:
              if key in keep:
                continue
            del graph.edata[key]
```

**Comment by jermainewang** [comment]
The introduction of `DGLGraph.local_scope` and `DGLGraph.local_var` might help this case. Close this issue since it's pretty old. Please reopen if the problem still exists.
